modules/objectdb_profiler/editor/objectdb_profiler_panel.cpp

```diff
@@ -0,0 +1,447 @@
+	TreeItem *item = snapshot_list->create_item(snapshot_list->get_root());
+	item->set_text(0, p_snapshot_file_name);
+	item->set_metadata(0, p_full_file_path);
+	item->move_before(snapshot_list->get_root()->get_first_child());
+	_update_diff_items();
+	return item;
+}
+
+void ObjectDBProfilerPanel::_show_selected_snapshot() {
+	if (snapshot_list->get_selected()->get_text(0) == diff_options[diff_button->get_selected_id()]) {
+		for (int i = 0; i < diff_button->get_item_count(); i++) {
+			if (diff_button->get_item_text(i) == current_snapshot->ptr()->name) {
+				diff_button->select(i);
+				break;
+			}
+		}
+	}
+	show_snapshot(snapshot_list->get_selected()->get_text(0), diff_options[diff_button->get_selected_id()]);
+	_update_enabled_diff_items();
+}
+
+Ref<GameStateSnapshotRef> ObjectDBProfilerPanel::get_snapshot(const String &p_snapshot_file_name) {
+	if (snapshot_cache.has(p_snapshot_file_name)) {
+		return snapshot_cache.get(p_snapshot_file_name);
+	} else {
+		Ref<DirAccess> snapshot_dir = _get_and_create_snapshot_storage_dir();
+		if (snapshot_dir.is_null()) {
+			ERR_PRINT("Could not access ObjectDB Snapshot directory");
+			return nullptr;
+		}
+
+		String full_file_path = snapshot_dir->get_current_dir().path_join(p_snapshot_file_name) + ".odb_snapshot";
+
+		Error err;
+		Ref<FileAccess> snapshot_file = FileAccess::open(full_file_path, FileAccess::READ, &err);
+		if (err != OK) {
+			ERR_PRINT("Could not open ObjectDB Snapshot file: " + full_file_path);
+			return nullptr;
+		}
+
+		Vector<uint8_t> content = snapshot_file->get_buffer(snapshot_file->get_length()); // we want to split on newlines, so normalize them
+		if (content.is_empty()) {
+			ERR_PRINT("ObjectDB Snapshot file is empty: " + full_file_path);
+			return nullptr;
+		}
+
+		Ref<GameStateSnapshotRef> snapshot = GameStateSnapshot::create_ref(p_snapshot_file_name, content);
+		if (snapshot.is_valid()) {
+			// don't cache a null snapshot
+			snapshot_cache.insert(p_snapshot_file_name, snapshot);
+		}
+		return snapshot;
+	}
+}
+
+void ObjectDBProfilerPanel::show_snapshot(const String &p_snapshot_file_name, const String &p_snapshot_diff_file_name) {
+	clear_snapshot();
+
+	current_snapshot = get_snapshot(p_snapshot_file_name);
+	if (p_snapshot_diff_file_name != "none") {
+		diff_snapshot = get_snapshot(p_snapshot_diff_file_name);
+	} else {
+		diff_snapshot.unref();
+	}
+
+	_view_tab_changed(view_tabs->get_current_tab());
+}
+
+void ObjectDBProfilerPanel::_view_tab_changed(int p_tab_idx) {
+	// populating tabs only on tab changed because we're handling a lot of data,
+	// and the editor freezes for while if we try to populate every tab at once
+	SnapshotView *view = cast_to<SnapshotView>(view_tabs->get_current_tab_control());
+	GameStateSnapshot *snapshot = current_snapshot.is_null() ? nullptr : current_snapshot->ptr();
+	GameStateSnapshot *diff = diff_snapshot.is_null() ? nullptr : diff_snapshot->ptr();
```

Comment:
```suggestion
	GameStateSnapshot *snapshot = current_snapshot->ptr();
	GameStateSnapshot *diff = diff_snapshot->ptr();
```

Reply:
This doesn't seem to work. I think it's because `current_snapshot` is a `Ref<GameStateSnapshotRef>`, which is a pointer wrapping a pointer, so I need to check if the outer pointer is null before reading the inner one.

If I'm using refs wrong in `Ref<GameStateSnapshotRef>`, please let me know.

Reply:
Oh my bad missed that it was `->ptr`